### ISSUE TITLE
Getting version errors while ucing tc-redirect-tap

### ISSUE THREAD

**Issue by AvaterClasher** [awslabs/tc-redirect-tap#98]
I am trying to setup a go project with firecrackers-go-sdk so for networking
I tried using cni version: v1.0.0 first for the plugin that gave me this error :

```
Failed handler "fcinit.SetupNetwork": failure when invoking CNI: failed to delete pre-existing CNI network {NetworkName:fcnet NetworkConfig:<nil> IfName:veth0 VMIfName: Args:[] BinPath:[/opt/cni/bin] ConfDir:/etc/cni/conf.d CacheDir:/var/lib/cni/b390d0c4-efd4-442b-96fe-d18af3c459af containerID:b390d0c4-efd4-442b-96fe-d18af3c459af netNSPath:/var/run/netns/b390d0c4-efd4-442b-96fe-d18af3c459af Force:false}: failed to delete CNI network list "fcnet": plugin type="tc-redirect-tap" failed (delete): incompatible CNI versions; config is "1.0.0", plugin supports ["0.3.0" "0.3.1" "0.4.0" "1.1.0"] 
```

So I tried changing the version to the suggedted v0.4.0 and that gives me this error now:

```
Failed handler "fcinit.SetupNetwork": failure when invoking CNI: failed to create CNI network: plugin type="tc-redirect-tap" failed (add): unsupported CNI result version "1.1.0"
```

The thing is here i am using this fcnet.conflist instead:

```
{
  "name": "fcnet",
  "cniVersion": "0.4.0",
  "plugins": [
    {
      "type": "ptp",
      "ipMasq": true,
      "ipam": {
        "type": "host-local",
        "subnet": "192.168.127.0/24",
        "resolvConf": "/etc/resolv.conf"
      }
    },
    {
      "type": "tc-redirect-tap"
    }
  ]
}
```

How should this be setup then ?

**Comment by austinvazquez** [comment]
@AvaterClasher, thanks for reporting the issue. This sounds like a bug to me as tc-redirect-tap does support version 1.0.0. I believe this is caused by the supported version logic depending on CNI `version.CurrentVersion` function which is currently returning `1.1.0` and missing `1.0.0` has broken the chain. I have opened a PR to fix the issue if you can please take a look.

**Comment by AvaterClasher** [comment]
@austinvazquez Hello, I tested the plugin with the fix you did but I am still getting the same error. I think I did everything correctly cloned your repo, checked out to your branch, rebuilt the plugin and put it in my `cni/bin`. 
This is the repo I used for checking if its spawning if it helps. [Repo](https://github.com/AvaterClasher/firecracker-test).
This is the current error. ->
```
2024/11/30 13:31:00 Attempting to spawn a Firecracker VM...
WARN[0001] Failed handler "fcinit.SetupNetwork": failure when invoking CNI: failed to create CNI network: plugin type="tc-redirect-tap" failed (add): unsupported CNI result version "1.1.0" 
2024/11/30 13:31:02 Failed to spawn Firecracker VM: failed to start machine: failure when invoking CNI: failed to create CNI network: plugin type="tc-redirect-tap" failed (add): unsupported CNI result version "1.1.0"
```
And the current conf ->
```
{
  "name": "fcnet",
  "cniVersion": "1.0.0",
  "plugins": [
    {
      "type": "ptp",
      "ipMasq": true,
      "ipam": {
        "type": "host-local",
        "subnet": "192.168.127.0/24",
        "resolvConf": "/etc/resolv.conf"
      }
    },
    {
      "type": "tc-redirect-tap"
    }
  ]
}
```

**Comment by Kern--** [comment]
Hi @AvaterClasher 

That later `ptp` issue looks like you might be using an older version of the ptp plugin. Looks like the standard cni plugins go support for CNI `1.1.0` in v1.6.0. Can you confirm that you are up to date? 

You can confirm with `/opt/cni/bin/ptp version` - it should report up to `1.1.0`.



For the earlier issue, I wonder if there's something weird happening because firecracker-go-sdk is using CNI 1.0.1 to exec the plugins. If you still want to try CNI 1.0.0, can you try building against the firecracker-go-sdk main branch?

**Comment by AvaterClasher** [comment]
Hi @Kern-- 
Checked the ptp version it indeed reports upto `1.1.0`.
Will try out what you suggested with the firecracker-go-sdk.

Another thing I tried out was hacking the `tc-redirect-tap` a little bit.

So I found out that this line was giving out the version to be `v1.1.0`. No matter what my cni version was.
https://github.com/awslabs/tc-redirect-tap/blob/c5566df386ae8533288438c475085655dfbf8257/cmd/tc-redirect-tap/main.go#L68

So I hardcoded it to `v1.0.0` and it started working as expected.
https://github.com/AvaterClasher/tc-redirect-tap/blob/321e88dc7aaf3ecda6fb945b4b91aeda040fbce5/cmd/tc-redirect-tap/main.go#L69

Now I dont know much about CNI plugins to tell what was going wrong. Hope it helps.

**Comment by sondavidb** [comment]
Hey, sorry for the delay in response.

TL;DR `go get github.com/containernetworking/cni@v1.2.0` (or later) with the new `tc-redirect-tap` built from the aforementioned PR should fix this issue.

I want to confirm that I was able to reproduce your issue. I cloned your repo with an empty rootfs in `./rootfs` with `tc-redirect-tap` installed and your exact config in `/etc/cni/conf.d/fcnet.conflist`, which gave me the same error when using `v0.4.0` and `v1.0.0` for the ptp plugin.

The problem was that the tc-redirect-tap CNI version being reported was `v1.1.0` which firecracker-go-sdk did not recognize as a valid version, since the release you were using is on v1.0.1 of `github.com/containernetworking/cni`. Support for v1.1.0 got added in v1.2.0 (specifically this [commit](https://github.com/containernetworking/cni/commit/a899051e444b287acfc637076d5f2226a834588b#diff-d77aab64b65fdf872ce9a0018b0d2922dd51b1638e0247be00dd084dd0d53b7eL29)). So when specifying CNI v1.1.0 it failed because it didn't recognize it.

It is definitely a little strange to me, though, that the line you mentioned always returns v1.1.0. It makes sense, but if it supports multiple CNI versions, and you specify v0.4.0, which is supported, it shouldn't error out. I'll have to dig deeper...

Anyway, feel free to let me know if this solves your issue. Thanks again!

**Comment by sondavidb** [comment]
TL;DR Opened a new PR to fix an adjacent-related problem. This PR as well as the previous merged PR should (hopefully!) completely resolve all of the issues stemmed from this.

Ok, I got back around to figuring out why it always returns v1.1.0.

When creating a plugin and adding it, we rely on the current result's CNI version to know which CNI version to return, but this was always populated by the implemented spec version, so in this case it would always return v1.1.0. The behavior was weird, but the implementation at a glance seemed fine.

So then I took a look at how ptp does it and I noticed that when they return [conf.CNIVersion](https://github.com/containernetworking/plugins/blob/main/plugins/main/ptp/ptp.go#L248), it was being taken directly from unmarshaling the JSON contents into their struct. Similar thing happens in the bridge plugin [here](https://github.com/containernetworking/plugins/blob/ba8bc7d0c70ab99d1a484982b437be92dc509e56/plugins/main/bridge/bridge.go#L745), where [cniVersion](https://github.com/containernetworking/plugins/blob/ba8bc7d0c70ab99d1a484982b437be92dc509e56/plugins/main/bridge/bridge.go#L522) is loaded from the configuration and given its own variable before being returned directly. Note that while both functions create a result and return it, the CNI version that's returned with it is separate from the new result that is created, because these results are (probably) going to be populated by the latest CNI version they implement.

Meanwhile, we use the CNI version [returned](https://github.com/awslabs/tc-redirect-tap/blob/f2644ce8d39d014db0f89e72f99a1c3710d7f041/cmd/tc-redirect-tap/main.go#L67) from the result (as you noticed above), which is always going to be v1.1.0.

Fixing this just meant we had to store the CNI version loaded from the JSON separately and then make sure we can reference it when returning from the add function.

Hopefully this should fully solve your use case, and please let us know if anything else comes up. Otherwise, if the above PR addresses your concerns, feel free to close the issue.

Thanks!

**Comment by AvaterClasher** [comment]
Hi @sondavidb,

Thank you so much for taking the time to dig into this issue and for the detailed explanation!

I'll be using Firecracker to build a cool college project, and your resolution has definitely made the process a lot smoother. I'll be sure to share updates once I have something interesting to showcase.

Thanks again for your help!
